src/main/java/uk/ac/ebi/eva/commons/core/models/VariantSourceEntryWithSamples.java

```
@@ -0,0 +1,95 @@
+        );
+    }
+
+    public VariantSourceEntryWithSamples(VariantSourceEntryMongo variantSourceEntryMongo, List<String> sampleNames) {
+        this(
+                variantSourceEntryMongo.getFileId(),
+                variantSourceEntryMongo.getStudyId(),
+                variantSourceEntryMongo.getSecondaryAlternates(),
+                variantSourceEntryMongo.getFormat(),
+                null,
+                variantSourceEntryMongo.getAttributes(),
+                joinSamplesDataWithSampleNames(variantSourceEntryMongo.deflateSamplesData(sampleNames.size()), sampleNames)
+        );
+    }
+
+    public VariantSourceEntryWithSamples(String fileId, String studyId, String[] secondaryAlternates, String format,
+                                         Map<String, VariantStatistics> cohortStats, Map<String, String> attributes,
+                                         LinkedHashMap<String, Map<String, String>> samplesData) {
+        super(fileId, studyId, secondaryAlternates, format, cohortStats, attributes);
+        this.samplesData = new LinkedHashMap<>();
+        if (samplesData != null) {
+            this.samplesData.putAll(samplesData);
+        }
+    }
+
+    public List<Map<String, String>> getSamplesData() {
+        return new ArrayList<>(samplesData.values());
+    }
+
+    public Map<String, Map<String, String>> getSamplesDataMap() {
+        return samplesData;
+    }
+
+    /**
+     * Joins the list of sample data with their correspondent name from the sample list
+     *
+     * @param samplesData
+     * @param samples
+     * @return
+     */
+    private static LinkedHashMap<String, Map<String, String>> joinSamplesDataWithSampleNames(
+            List<Map<String, String>> samplesData,
+            List<String> samples) {
+        LinkedHashMap<String, Map<String, String>> temp = new LinkedHashMap<>();
+        int numberOfSamples = Math.min(samples.size(), samplesData.size());
```

I would prefer getting an ArrayIndexOutOfBondExceptions that getting a silent error if the samples and samplesData lists are not syncronized. We could be storing incorrect data without knowing it.